keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry/README.md

```diff
@@ -0,0 +1,545 @@
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+Alice is setting up a development environment on Kubernetes.
+
+Alice creates a local Kind cluster. She follows the Kind-specific instructions
+for setting up a local registry during cluster creation.
+
+Alice will have some tool she interacts with that builds images and pushes them
+to a registry the cluster can access. That tool might be an IDE like VSCode,
+an infra-as-code toolchain like Pulumi, or a multi-service dev environment like Tilt.
+
+On startup, the tool should connect to the cluster and read the registry ConfigMap.
+
+If the ConfigMap specifies a `help` URL, the tool may prompt the user to set up
+a registry for faster development.
+
+If the ConfigMap specifies a registry location, the tool may automatically
+adjust image pushes to push to the specified registry.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+- Users may see the ConfigMap and draw mistaken conclusions about how to
+  interact with it. They think deleting the ConfigMap would delete the local
+  registry (which it does not).
+
+- This KEP does not specify any automatic mechanism for keeping the
+  RegistryStatus up-to-date with how the cluster is configured.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+Tools that configure a local registry should apply the following ConfigMap to
+the cluster:
+
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: registry
+  namespace: kube-public
+data:
+  RegistryStatus: |
+    apiVersion: registry.x-k8s.io/v1alpha1
+    kind: RegistryStatus
+    host: "localhost:5000"
+    hostFromCluster: "registry:5000"
+    help: "https://kind.sigs.k8s.io/docs/user/local-registry/"
+```
+
+Documentation that educates people on how to set up a registry manually should
+include instructions on this.
+
+The name of the config map must be `registry`.
+
+The namespace must be `kube-public`.
+
+The `RegistryStatus` field of the ConfigMap data takes the form of a component
+config, like the `kubeadm-config` ConfigMap in `kube-system`.
```

Comment:
one difference with the `kubeadm-config` CM is that kubeadm is supposedly the only machine-reader/writer of the CM in question.

on one hand, component config (and versioned config in general), opens a bit of a can of worms:
- has to graduate from alpha->v1 eventually and the SIG needs to track this work with the owner of the API / KEP.
- needs to have it's conversion machinery reside in kubernetes-sig ORG repository owned by the SIG.
- readers of the ConfigMap need to consume the API machinery as code, so that can convert between the version in the ConfigMap and the latest preferred version.
- writers of the ConfigMap need to consume only the publicly exposed API type without the conversion logic.
- if a cluster deployer writes a ConfigMap of version v1alpha1 that was deprecated, but the machinery of the ConfigMap reader have already removed support for this API the conversion is expected to fail.
- we also have the problem of how to do conversions. the standard API machinery in k8s has public/private types. private types are required for conversion (kubeadm uses that). kind on the other hand implements its own [conversion logic](https://github.com/kubernetes-sigs/kind/tree/master/pkg/internal/apis/config), which is arguably simpler.

on the other hand, non-versioned config breaks users if fields are deleted/renamed and still requires machinery for transitions.

Reply:
These are all great questions. I struggled with this a lot while considering this approach.

The fundamental problem (I think?) is that the k8s api machinery is built for a world where you can assume all the consumers of the API go through a central codepath. e.g., they read objects through the api server. Or there's a single reader. There's a natural "gatekeeper" for conversion and governance.

In this case, there are multiple readers and writers who don't want to pull in the apimachinery, and don't have any central governance.

I'm not sure if there's a way to slice this pizza in a way that would satisfy all stakeholders. I'm curious about Kind-team's experience with how it does API types, and if you all would recommend that approach (cc @BenTheElder ).

I do not know how much buy-in we would need to create a kubernetes-sig org repository that holds a reference implementation of the API reader, or if that approach makes sense.

Reply:
To be clear, @neolit123, it sounds like you don't have a strong sense yet of whether this config should be unversioned or versioned -- you're still just thinking through the implications. is that right?

I don't know much about the typical development path of versioned apis, and what decisions need to be made when

Reply:
> To be clear, @neolit123, it sounds like you don't have a strong sense yet of whether this config should be unversioned or versioned -- you're still just thinking through the implications. is that right?

yes, just evaluating the cost vs benefits.

> The fundamental problem (I think?) is that the k8s api machinery is built for a world where you can assume all the consumers of the API go through a central codepath. e.g., they read objects through the api server. Or there's a single reader. There's a natural "gatekeeper" for conversion and governance.
> In this case, there are multiple readers and writers who don't want to pull in the apimachinery, and don't have any central governance.

yes, that is true. for component config we faced some of these issues.

> I'm not sure if there's a way to slice this pizza in a way that would satisfy all stakeholders.

the main question to answer here is how do we expect the API to change. if it's not going to change, not-versioning it should be fine. yet that can bite down the line.

> I do not know how much buy-in we would need to create a kubernetes-sig org repository that holds a reference implementation of the API reader, or if that approach makes sense.

the creation and management of the repository can be done, but it definitely falls heavily under "costs". it has benefits, but for something that is not going to be fast moving (e.g. it might remain at 3 fields for a while) it's hard to justify a dedicated repository.